### PR TITLE
WIP: Initial commit of adding build discipline

### PR DESCRIPTION
**Pull request by helena**
In progress: refining, testing against scala and java versions, testing all tasks #26206

### CODE REVIEW

File: project/Versions.scala [hunk 1]
```diff
@@ -0,0 +1,125 @@
+package akka
+
+import sbt._
+import Keys._
+
+object Versions {
```
Comment (helena) [inline project/Versions.scala]: * If we want, we can link this in docs for referencing versions if needed
* Makes it very clean and fast to upgrade versions
* Almost all version-related logic is now here, where makes sense

File: project/Versions.scala [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    Seq(
+      crossScalaVersions := Scala,
+      scalaVersion := sys.props.getOrElse("akka.build.scalaVersion", crossScalaVersions.value.head),
+      scalaStmVersion := sys.props.getOrElse("akka.build.scalaStmVersion", StmDefault),
+      //libraryDependencies += "org.scalacheck" %% "scalacheck" % scalaCheck(scalaVersion.value)
+      scalaCheckVersion := sys.props.getOrElse("akka.build.scalaCheckVersion", scalaCheck(scalaVersion.value)), // unmodified 3-clause BSD
+      scalaTestVersion := ScalaTest,
+      java8CompatVersion := java8Compat(scalaVersion.value))
+  }
+
```
Comment (helena) [inline project/Versions.scala]: 'notOnScala211' and 'notOnJdk8' are from Jdk9.scala

File: project/Jdk9.scala [hunk 3]
```diff
@@ -8,18 +8,10 @@ import sbt._
 import sbt.Keys._
 
 object Jdk9 extends AutoPlugin {
+  import akka.Versions.{notOnJdk8, notOnScala211}
 
   lazy val CompileJdk9 = config("CompileJdk9").extend(Compile)
```
Comment (helena) [inline project/Jdk9.scala]: These 2 functions moved to Versions.scala

File: project/Jdk9.scala [hunk 4]
```diff
@@ -8,18 +8,10 @@ import sbt._
...
 
-  def notOnScala211[T](scalaBinaryVersion: String, values: Seq[T]): Seq[T] = scalaBinaryVersion match {
-    case "2.11" => Seq()
-    case _ => values
-  }
```
Comment (helena) [inline project/Jdk9.scala]: moved to Versions.scala

File: project/Jdk9.scala [hunk 5]
```diff
@@ -8,18 +8,10 @@ import sbt._
...
-
-  def notOnJdk8[T](values: Seq[T]): Seq[T] =
-    if (System.getProperty("java.version").startsWith("1.")) Seq()
-    else values
```
Comment (helena) [inline project/Jdk9.scala]: moved to Versions.scala

File: project/AkkaDisciplinePlugin.scala [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    if (isScalaMinor(">=", 13, scalaVersion))
+      Seq("-Ymacro-annotations")         // Enable support for macro annotations, formerly in macro paradise.
+    else
+      Nil
+
+  // TODO ?
+  lazy val flagsNoAutoImports = Seq(
+    "-Yno-imports"                       // No automatic imports at all; all symbols must be imported explicitly
+  )
+
```
Comment (helena) [inline project/AkkaDisciplinePlugin.scala]: Will eventually remove the commented settings when fully tested.

File: project/AkkaDisciplinePlugin.scala [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        Nil
+    }
+
+  private def additionsFor(scalaVersion: String): Seq[String] =
+    if (isScalaMinor(">=", 13, scalaVersion))
+      Seq("-Ymacro-annotations")         // Enable support for macro annotations, formerly in macro paradise.
+    else
+      Nil
+
+  // TODO ?
```
Comment (helena) [inline project/AkkaDisciplinePlugin.scala]: Testing:

File: project/AkkaDisciplinePlugin.scala [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+      Wart.OptionPartial,
+      Wart.Option2Iterable,
+      Wart.ArrayEquals,
+      Wart.ToString,
+      Wart.StringPlusAny,
+      Wart.While,
+      Wart.Product,
+      Wart.JavaSerializable,
+      Wart.Serializable),
+    wartremoverErrors in Test ++= Nil) // in progress
```
Comment (helena) [inline project/AkkaDisciplinePlugin.scala]: Not refined yet and so far only compiled against Akka-actor. Will finish tomorrow.

File: project/AkkaDisciplinePlugin.scala [hunk 9]
```diff
@@ -0,0 +1,191 @@
+package akka
+
+import sbt._, Keys._
+
+object CommonOptions {
```
Comment (helena) [inline project/AkkaDisciplinePlugin.scala]: For any modules not wanting discipline, though I hope eventually to apply to all, these common settings can just be used. It is almost exactly what is in master, with minimal change.

File: project/AkkaBuild.scala [hunk 10]
```diff
@@ -81,97 +81,85 @@ object AkkaBuild {
...
+
+  val testSettings =
+    TestExtras.Filter.settings ++ Seq(
       fork in Test := true,
 
+      scalacOptions in Test := (scalacOptions in Test).value.filterNot(opt ⇒
+        opt == "-Xlog-reflective-calls" || opt.contains("genjavadoc")) ++ Seq("-Ywarn-unused"),
+
+      javacOptions in test ++= CommonOptions.commonJavacOptions ++ CommonOptions.javaSource(
+        CrossJava.Keys.fullJavaHomes.value("8") + "/jre/lib/rt.jar"),
```
Comment (helena) [inline project/AkkaBuild.scala]: Ideally we can get this as one, e.g. from CrossJava, vs repeated but I'd need to set it up as a setting or task.

File: project/AkkaBuild.scala [hunk 11]
```diff
@@ -81,97 +81,85 @@ object AkkaBuild {
...
-      // Adds a `src/main/scala-2.13+` source directory for Scala 2.13 and newer
-      // and a `src/main/scala-2.13-` source directory for Scala version older than 2.13
-      unmanagedSourceDirectories in Compile += {
-        val sourceDir = (sourceDirectory in Compile).value
-        CrossVersion.partialVersion(scalaVersion.value) match {
-          case Some((2, n)) if n >= 13 => sourceDir / "scala-2.13+"
-          case _                       => sourceDir / "scala-2.13-"
-        }
-      },
+  // Probably going away with: https://github.com/akka/akka/issues/26088
```
Comment (helena) [inline project/AkkaBuild.scala]: Or not

File: project/Versions.scala (hunk 1 above)
Comment (patriknw) [reply project/Versions.scala]: When I want to check dependencies and versions of an unfamiliar project I look for `Dependencies.scala`. Should we stick to the convention of having these in Dependencies.scala?

For docs we can use https://github.com/lightbend/sbt-paradox-dependencies, which was developed and used by Alpakka team.

**Review by patriknw — COMMENTED** [review]
looking good, difficult to see what actually changed and what is just moved around. Could you give a summary of the important settings that changed?

**Comment by jrudolph** [comment]
> looking good, difficult to see what actually changed and what is just moved around. Could you give a summary of the important settings that changed?

Indeed, would be good to break down into several commits / PRs to ease and speed up reviewing.

File: project/Versions.scala (hunk 1 above)
Comment (helena) [reply project/Versions.scala]: * WRT convention, I've found more projects handle versions for dependencies in a Version(s) object. In Akka we have to name it Versions. 
* Having the configuration of dependencies mixed in with the versions is not as fast for upgrades and not as clean 

Generally I find it messy and better to compartmentalize this way.

**Comment by helena** [comment]
@patriknw  @jrudolph Sure. To clarify, it's cleanup, compartmentalizing for ease in maintenance and logic of build work, but **primarily** the addition of the new discipline plugin to the build, and it's additional compiler flags, error catching and linting to catch in compile time for all contributions, and the addition of scoverage. I will resubmit with separate, descriptive commits if it helps.

File: project/Versions.scala (hunk 1 above)
Comment (helena) [reply project/Versions.scala]: ok

**Comment by helena** [comment]
Closing in favor of https://github.com/akka/akka/pull/26376 with breakouts for clarity.

**Pull request closed without merging**